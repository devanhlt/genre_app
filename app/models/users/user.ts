import { Instance, SnapshotIn, SnapshotOut, types } from "mobx-state-tree"

import { withSetPropAction } from "../helpers/withSetPropAction"

export const UserModel = types
  .model("AccountModel")
  .props({
    active: types.optional(types.maybeNull(types.boolean), false),
    admin: types.optional(types.maybeNull(types.boolean), false),
    createdBy: types.optional(types.maybeNull(types.string), null),
    createdDate: types.optional(types.maybeNull(types.number), null),
    department: types.optional(types.maybeNull(types.string), null),
    departmentLite: types.optional(types.maybeNull(types.string), null),
    email: types.optional(types.maybeNull(types.string), null),
    fullName: types.optional(types.maybeNull(types.string), null),
    mobile: types.optional(types.maybeNull(types.string), null),
    status: types.optional(types.maybeNull(types.string), null),
    updatedBy: types.optional(types.maybeNull(types.string), null),
    updatedDate: types.optional(types.maybeNull(types.number), null),
    userName: types.optional(types.maybeNull(types.string), null),
    id: types.optional(
      types.maybeNull(
        types.model({
          date: types.optional(types.maybeNull(types.number), null),
          timestamp: types.optional(types.maybeNull(types.number), null),
        }),
      ),
      {
        date: null,
        timestamp: null,
      },
    ),
  })
  .actions(withSetPropAction)
  .views(() => ({}))

export interface User extends Instance<typeof UserModel> {}
export interface UserSnapshotOut extends SnapshotOut<typeof UserModel> {}
export interface UserSnapshotIn extends SnapshotIn<typeof UserModel> {}
