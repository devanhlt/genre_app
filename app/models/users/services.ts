import { ApiResponse } from "apisauce"

import { ApiServices } from "app/services/api"
import { GeneralApiProblem, getGeneralApiProblem } from "app/services/api/apiProblem"
import { User } from "./user"

export class UsersServices extends ApiServices {
  /**
   * Gets a list of system
   */
  async getUsers(): Promise<{ kind: "ok"; users: User[] } | GeneralApiProblem> {
    // make the api call
    const timestamp = new Date().getTime()
    const response: ApiResponse<User[]> = await this.apisauce.get(`api/v1/users?_=${timestamp}`)

    // the typical ways to die when calling an api
    if (!response.ok) {
      const problem = getGeneralApiProblem(response)
      if (problem) return problem
    }

    // transform the data into the format we are expecting
    try {
      const rawData = response.data

      // This is where we transform the data into the shape we expect for our MST model.
      const users: User[] = rawData.map((raw) => ({
        ...raw,
      }))

      return { kind: "ok", users }
    } catch (e) {
      if (__DEV__) {
        console.error(`Bad data: ${e.message}\n${response.data}`, e.stack)
      }
      return { kind: "bad-data" }
    }
  }
}
