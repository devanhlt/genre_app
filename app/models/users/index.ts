import { Instance, SnapshotOut, flow, getType, toGenerator, types } from "mobx-state-tree"

import { withEnvironment } from "../extensions/with-environment"
import { withRootStore } from "../helpers/withRootStore"
import { withSetPropAction } from "../helpers/withSetPropAction"
import { UsersServices } from "./users.services"
import { UserModel } from "./user"

export const UsersStoreModel = types
  .model("SystemStore")
  .props({
    users: types.optional(types.maybeNull(types.array(UserModel)), []),
  })
  .extend(withEnvironment) // Extend environment
  .extend(withRootStore)
  .actions(withSetPropAction)
  .views((self) => {
    return {
      get usersForList() {
        return self.users
      },
    }
  })
  .actions((self) => ({
    fetchAccounts: flow(function* fetchAccounts() {
      const usersServices = new UsersServices(self.rootStore)
      const response = yield* toGenerator(usersServices.getUsers())
      return response.kind === "ok" ? response.users : []
    }),

    /**
     * Reset model
     */
    reset: () => {
      self.setProp("users", [])
    },

    /**
     * Handle logout
     * Reset store
     */
    // Call when Instantiated
    afterCreate() {
      console.log("Instantiated " + getType(self).name)
    },
  }))

export interface UsersStore extends Instance<typeof UsersStoreModel> {}
export interface UsersStoreSnapshot extends SnapshotOut<typeof UsersStoreModel> {}
