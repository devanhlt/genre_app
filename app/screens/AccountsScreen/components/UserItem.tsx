import React from "react"
import { TextStyle, View, ViewStyle } from "react-native"
import Swipeable from "react-native-gesture-handler/Swipeable"

import { Typography } from "app/components/Typography"
import { Divider } from "app/components/Divider"
import { If } from "app/components/If"
import { Pressable } from "app/components/Pressable"
import { User } from "app/models/users/user"
import { appColors, spacing } from "app/theme"
import { responsiveWidth } from "app/utils/screens"

export interface UserItemProps {
  item?: User
  onPressItem?: (item: User) => void
}

export default function UserItem({ item }: UserItemProps): JSX.Element {
  if (!item) return null

  return (
    <Swipeable key={`user-item-${item.userName}`} renderRightActions={LeftSwipeActions}>
      <View style={$viewContainer}>
        <View style={$rowItem}>
          <Typography text="Full name:" preset="body03" style={$titleText} />
          <Divider size={spacing.size12} type="vertical" />
          <Typography text={item.fullName} preset="body03" style={$valueText} />
        </View>
        <View style={$rowItem}>
          <Typography text="Username:" preset="support01" style={$titleText} />
          <Divider size={spacing.size12} type="vertical" />
          <Typography text={item.userName} preset="support01" style={$valueText} />
        </View>
        <If condition={!!item.email}>
          <View style={$rowItem}>
            <Typography text="Email:" preset="support01" style={$titleText} />
            <Divider size={spacing.size12} type="vertical" />
            <Typography text={item.email} preset="support01" style={$valueText} />
          </View>
        </If>
      </View>
    </Swipeable>
  )
}

const LeftSwipeActions = () => {
  return (
    <View style={$leftSwipeActions}>
      <Pressable style={$actionButtonAllow}>
        <Typography
          text={"Allow admin"}
          preset="body02"
          color={appColors.palette.neutral0}
          style={$valueText}
        />
      </Pressable>
      <Pressable style={$actionButtonActive}>
        <Typography
          text={"Active"}
          preset="body02"
          color={appColors.palette.neutral0}
          style={$valueText}
        />
      </Pressable>
    </View>
  )
}

const $viewContainer: ViewStyle = {
  borderBottomWidth: 1,
  borderBottomColor: appColors.common.borderDivider,
  paddingVertical: spacing.size08,
  backgroundColor: appColors.common.bgWhite,
  paddingHorizontal: spacing.size16,
}

const $leftSwipeActions: ViewStyle = {
  flexDirection: "row",
  backgroundColor: appColors.palette.neutral300,
}

const $actionButtonAllow: ViewStyle = {
  backgroundColor: appColors.palette.neutral400,
  alignItems: "center",
  justifyContent: "center",
}

const $actionButtonActive: ViewStyle = {
  backgroundColor: appColors.palette.green300,
  alignItems: "center",
  justifyContent: "center",
}

const $titleText: TextStyle = {
  minWidth: responsiveWidth(75),
}

const $valueText: TextStyle = {
  flexWrap: "wrap",
  overflow: "hidden",
}

const $rowItem: ViewStyle = {
  flex: 1,
  flexDirection: "row",
  overflow: "hidden",
}
