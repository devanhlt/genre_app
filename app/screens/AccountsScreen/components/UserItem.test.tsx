import React from "react"

import renderer from "react-test-renderer"
import UserItem from "./UserItem"
import { User } from "app/models/users/user"

jest.mock("react-native-gesture-handler", () => undefined)

describe("<UserItem />", () => {
  it("renders correctly", () => {
    const user = {
      active: false,
      admin: false,
      createdBy: "Test",
      createdDate: new Date().getTime(),
      department: "TEST",
      departmentLite: "TEST",
      email: "test@gmail.com",
      fullName: "FULL NAME",
      mobile: "09xxxxxxxx",
      status: "active",
      updatedBy: "Test",
      updatedDate: new Date().getTime(),
      userName: "GVL00000",
      id: null,
    } as User
    const tree = renderer.create(<UserItem item={user} />)
    expect(tree).toMatchSnapshot()
  })
})
