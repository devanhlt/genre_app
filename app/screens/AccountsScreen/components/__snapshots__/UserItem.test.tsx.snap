// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserItem /> renders correctly when item data is correctly 1`] = `
<View
  collapsable={false}
  forwardedRef={[Function]}
  handlerTag={1}
  handlerType="PanGestureHandler"
  onGestureHandlerEvent={[Function]}
  onGestureHandlerStateChange={[Function]}
  onLayout={[Function]}
  style={
    {
      "overflow": "hidden",
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "bottom": 0,
        "flexDirection": "row-reverse",
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": [
          {
            "translateX": -10000,
          },
        ],
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#BBC3CF",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "paddingHorizontal": 16,
                "paddingVertical": 12,
              },
              {
                "backgroundColor": "#C5281C",
                "paddingHorizontal": 12,
                "paddingVertical": 12,
              },
            ],
            {
              "alignItems": "center",
              "backgroundColor": "#A1ABBB",
              "justifyContent": "center",
            },
            false,
            false,
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "flexGrow": 0,
                  "flexShrink": 1,
                  "marginStart": 4,
                  "textAlign": "center",
                  "zIndex": 2,
                },
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "600",
                  "letterSpacing": 0.15,
                  "lineHeight": 20,
                },
                {
                  "color": "#FFFFFF",
                },
              ],
              undefined,
              false,
              false,
            ]
          }
        >
          <Text
            preset="body02"
            style={
              [
                {},
                {
                  "color": "#FFFFFF",
                },
                {
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "letterSpacing": 0.15,
                  "lineHeight": 24,
                },
                {
                  "flexWrap": "wrap",
                  "overflow": "hidden",
                },
              ]
            }
          >
            Allow admin
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          [
            [
              {
                "paddingHorizontal": 16,
                "paddingVertical": 12,
              },
              {
                "backgroundColor": "#C5281C",
                "paddingHorizontal": 12,
                "paddingVertical": 12,
              },
            ],
            {
              "alignItems": "center",
              "backgroundColor": "#339633",
              "justifyContent": "center",
            },
            false,
            false,
          ]
        }
      >
        <View
          style={
            [
              [
                {
                  "flexGrow": 0,
                  "flexShrink": 1,
                  "marginStart": 4,
                  "textAlign": "center",
                  "zIndex": 2,
                },
                {
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "600",
                  "letterSpacing": 0.15,
                  "lineHeight": 20,
                },
                {
                  "color": "#FFFFFF",
                },
              ],
              undefined,
              false,
              false,
            ]
          }
        >
          <Text
            preset="body02"
            style={
              [
                {},
                {
                  "color": "#FFFFFF",
                },
                {
                  "fontFamily": "System",
                  "fontSize": 16,
                  "fontWeight": "400",
                  "letterSpacing": 0.15,
                  "lineHeight": 24,
                },
                {
                  "flexWrap": "wrap",
                  "overflow": "hidden",
                },
              ]
            }
          >
            Active
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
    />
  </View>
  <View
    collapsable={false}
    forwardedRef={[Function]}
    handlerTag={2}
    handlerType="TapGestureHandler"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    pointerEvents="auto"
    style={
      {
        "transform": [
          {
            "translateX": -0,
          },
        ],
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#FFFFFF",
          "borderBottomColor": "#E5E8ED",
          "borderBottomWidth": 1,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        }
      }
    >
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "overflow": "hidden",
          }
        }
      >
        <Text
          preset="body03"
          style={
            [
              {},
              {
                "color": "#000F27",
              },
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "600",
                "letterSpacing": 0.15,
                "lineHeight": 20,
              },
              {
                "minWidth": 75,
              },
            ]
          }
        >
          Full name:
        </Text>
        <View
          style={
            [
              {
                "flexGrow": 0,
                "flexShrink": 0,
              },
              false,
              {
                "width": 12,
              },
              undefined,
            ]
          }
        />
        <Text
          preset="body03"
          style={
            [
              {},
              {
                "color": "#000F27",
              },
              {
                "fontFamily": "System",
                "fontSize": 14,
                "fontWeight": "600",
                "letterSpacing": 0.15,
                "lineHeight": 20,
              },
              {
                "flexWrap": "wrap",
                "overflow": "hidden",
              },
            ]
          }
        >
          FULL NAME
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "overflow": "hidden",
          }
        }
      >
        <Text
          preset="support01"
          style={
            [
              {},
              {
                "color": "#000F27",
              },
              {
                "fontFamily": "System",
                "fontSize": 12,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 16,
              },
              {
                "minWidth": 75,
              },
            ]
          }
        >
          Username:
        </Text>
        <View
          style={
            [
              {
                "flexGrow": 0,
                "flexShrink": 0,
              },
              false,
              {
                "width": 12,
              },
              undefined,
            ]
          }
        />
        <Text
          preset="support01"
          style={
            [
              {},
              {
                "color": "#000F27",
              },
              {
                "fontFamily": "System",
                "fontSize": 12,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 16,
              },
              {
                "flexWrap": "wrap",
                "overflow": "hidden",
              },
            ]
          }
        >
          GVL00000
        </Text>
      </View>
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "overflow": "hidden",
          }
        }
      >
        <Text
          preset="support01"
          style={
            [
              {},
              {
                "color": "#000F27",
              },
              {
                "fontFamily": "System",
                "fontSize": 12,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 16,
              },
              {
                "minWidth": 75,
              },
            ]
          }
        >
          Email:
        </Text>
        <View
          style={
            [
              {
                "flexGrow": 0,
                "flexShrink": 0,
              },
              false,
              {
                "width": 12,
              },
              undefined,
            ]
          }
        />
        <Text
          preset="support01"
          style={
            [
              {},
              {
                "color": "#000F27",
              },
              {
                "fontFamily": "System",
                "fontSize": 12,
                "fontWeight": "400",
                "letterSpacing": 0.25,
                "lineHeight": 16,
              },
              {
                "flexWrap": "wrap",
                "overflow": "hidden",
              },
            ]
          }
        >
          test@gmail.com
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<UserItem /> renders correctly when item is empty 1`] = `null`;
